File: p3/p338/CountBits.java
package leetcode.p3.p338;

import java.util.Arrays;

public class CountBits {
    /**
     * 09/05/2022 20:11
     * My approach
     * DP and change the leftmost 1 bit to 0
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public static int[] countBits(int n) {
        int[] result = new int[n + 1];
        result[0] = 0;
        if (n == 0) {
            return result;
        }
        result[1] = 1;
        for (int i = 2; i <= n; i++) {
            for (int j = 0; j < 32; j++) {
                if (((1 << 31 - j) & i) == 1 << 31 - j) {
                    result[i] = result[i & ~(1 << 31 - j)] + 1;
                    break;
                }
            }
        }
        return result;
    }

    // TODO

    public static void main(String[] args) {
        System.out.println(Arrays.toString(countBits(5)));
    }
}

File: p3/p338/count_bits.py
from typing import List


class Solution:
    """
    Feb 19, 2025 15:11
    Time Complexity: O(nlogn)
    Space Complexity: O(1)
    """

    def countBits(self, n: int) -> List[int]:
        ans = []
        for i in range(n + 1):
            ans.append(bin(i).count("1"))

        return ans


class Solution1:
    """
    Feb 19, 2025 15:30
    Time Complexity: O(n)
    Space Complexity: O(1)
    """

    def countBits(self, n: int) -> List[int]:
        ans = [0]
        if n == 0:
            return ans

        for i in range(1, n + 1):
            ans.append(ans[i >> 1] + (i & 1))
        return ans
